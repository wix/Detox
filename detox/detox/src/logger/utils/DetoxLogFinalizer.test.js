jest.unmock('../DetoxLogger');

const path = require('path');
const { PassThrough } = require('stream');

const fs = require('fs-extra');
const glob = require('glob');
const tempfile = require('tempfile');

const temporary = require('../../artifacts/utils/temporaryPath');
const DetoxLogger = require('../DetoxLogger');

const DetoxLogFinalizer = require('./DetoxLogFinalizer');

describe('DetoxLogFinalizer', () => {
  /** @type {string[]} */
  let temporaryFiles = [];
  /** @type {*} */
  let session;
  /** @type {import('./DetoxLogFinalizer')} */
  let finalizer;
  /** @type {import('../DetoxLogger')} */
  let fakeLogger;

  let _env;

  beforeEach(() => {
    _env = process.env;
    process.env = { ..._env, FORCE_COLOR: '0' };
  });

  beforeEach(() => {
    session = {
      id: 'session-id',
      detoxConfig: {
        artifacts: {
          rootDir: tempfile(''),
          plugins: {
            log: {
              enabled: true,
              keepOnlyFailedTestsArtifacts: false,
            }
          }
        },
        logger: {
          level: 'info',
          overrideConsole: false,
          options: {
            ...DetoxLogger.defaultOptions({
              level: 'trace',
            }),
            showDate: (date) => date.toISOString(),
            colors: false,
          },
        }
      }
    };

    temporaryFiles = [session.detoxConfig.artifacts.rootDir];

    const FakeLogger = jest.requireMock('../DetoxLogger');
    fakeLogger = new FakeLogger();

    finalizer = new DetoxLogFinalizer({ session, logger: fakeLogger });
  });

  afterEach(async () => {
    await Promise.all(temporaryFiles.map(f => fs.remove(f)));
  });

  afterEach(() => {
    process.env = _env;
  });

  describe('finalize', () => {
    it('should not throw when there are no logs', async () => {
      await finalizer.finalize();
      expect(fs.existsSync(artifactsDir())).toBe(false);
    });

    it('should convert JSONL logs to Chrome Trace format', async () => {
      await createLogFiles();
      await finalizer.finalize();

      let plainLog = await fs.readFile(path.join(artifactsDir(), 'detox.log'), 'utf8');
      let traceLog = JSON.parse(await fs.readFile(path.join(artifactsDir(), 'detox.trace.json'), 'utf8'));

      expect(plainLog).toMatchSnapshot('plain');
      expect(traceLog).toMatchSnapshot('chrome-trace');
    });

    it('should not create logs when they are disabled', async () => {
      session.detoxConfig.artifacts.plugins.log.enabled = false;
      await createLogFiles();
      await finalizer.finalize();
      expect(fs.existsSync(artifactsDir())).toBe(false);
      expect(fs.existsSync(temporaryFiles[1])).toBe(false);
      expect(fs.existsSync(temporaryFiles[2])).toBe(false);
    });

    it('should not create logs for successful run when configured', async () => {
      session.detoxConfig.artifacts.plugins.log.keepOnlyFailedTestsArtifacts = true;
      session.testResults = [{ success: true }];

      await createLogFiles();
      await finalizer.finalize();
      expect(fs.existsSync(artifactsDir())).toBe(false);
      expect(fs.existsSync(temporaryFiles[1])).toBe(false);
      expect(fs.existsSync(temporaryFiles[2])).toBe(false);
    });

    it('should create logs for failing run when configured', async () => {
      session.detoxConfig.artifacts.plugins.log.keepOnlyFailedTestsArtifacts = true;
      session.testResults = [{ success: false }];

      await createLogFiles();
      await finalizer.finalize();

      const outFiles = glob.sync('*', { cwd: artifactsDir() });
      expect(outFiles).toEqual([
        'detox.log',
        'detox.trace.json',
      ]);
    });
  });

  describe('finalizeSync', () => {
    it('should not throw for non-initialized state', async () => {
      await createLogFiles();
      delete session.detoxConfig.artifacts.plugins;
      expect(() => finalizer.finalizeSync()).not.toThrow();

      await createLogFiles();
      delete session.detoxConfig.artifacts.rootDir;
      expect(() => finalizer.finalizeSync()).not.toThrow();

      await createLogFiles();
      delete session.detoxConfig;
      expect(() => finalizer.finalizeSync()).not.toThrow();
    });

    it('should not throw when there are no logs', async () => {
      finalizer.finalizeSync();
      expect(fs.existsSync(artifactsDir())).toBe(false);
    });

    it('should copy JSONL logs to artifacts dir', async () => {
      await createLogFiles();
      finalizer.finalizeSync();

      const outFiles = glob.sync('*', { cwd: artifactsDir() });
      expect(outFiles.map(f => path.extname(f))).toEqual([
        '.jsonl',
        '.jsonl',
      ]);
    });

    it('should not create logs when they are disabled', async () => {
      session.detoxConfig.artifacts.plugins.log.enabled = false;
      await createLogFiles();
      finalizer.finalizeSync();
      expect(fs.existsSync(artifactsDir())).toBe(false);
      expect(fs.existsSync(temporaryFiles[1])).toBe(false);
      expect(fs.existsSync(temporaryFiles[2])).toBe(false);
    });

    it('should create logs even for successful run when configured to keep failing logs', async () => {
      session.detoxConfig.artifacts.plugins.log.keepOnlyFailedTestsArtifacts = true;
      session.testResults = [{ success: true }];

      await createLogFiles();
      finalizer.finalizeSync();

      const outFiles = glob.sync('*', { cwd: artifactsDir() });
      expect(outFiles.map(f => path.extname(f))).toEqual([
        '.jsonl',
        '.jsonl',
      ]);

      // assert that the temporary files have been moved, not copied
      expect(fs.existsSync(temporaryFiles[1])).toBe(false);
      expect(fs.existsSync(temporaryFiles[2])).toBe(false);
    });
  });

  describe('createEventStream', () => {
    it('should create a stream of Chrome Trace format events', async () => {
      await createLogFiles();

      const events = [];
      await new Promise((resolve, reject) => {
        finalizer.createEventStream()
          .on('end', resolve)
          .on('error', reject)
          .on('data', event => events.push(event));
      });

      events.forEach(t => { t.pid = 1234; });

      expect(events).toMatchSnapshot();
    });
  });

  function artifactsDir() {
    return session.detoxConfig.artifacts.rootDir;
  }

  async function createLogFiles() {
    const sessionId = session.id;
    const logPath1 = temporary.for.jsonl(`${sessionId}.1`);
    const logPath2 = temporary.for.jsonl(`${sessionId}.2`);

    temporaryFiles.push(logPath1, logPath2);

    const userConfig = {
      ...session.detoxConfig.logger,
      options: {
        ...session.detoxConfig.logger.options,
        out: new PassThrough(),
      },
    };

    const root1 = new DetoxLogger({
      unsafeMode: true,
      file: logPath1,
      userConfig,
    });

    const root2 = new DetoxLogger({
      unsafeMode: true,
      file: logPath2,
      userConfig,
    });

    const log1 = root1.child({ pid: 1234 });
    const log2 = root2.child({ pid: 1235 });

    jest.useFakeTimers();

    jest.setSystemTime(new Date('2023-01-01T00:00:00.000Z'));
    log1.info({ cat: 'main' }, 'msg1');

    jest.setSystemTime(new Date('2023-01-01T00:00:01.000Z'));
    log1.trace.begin({ id: 1, cat: 'parallel' }, 'duration1');

    jest.setSystemTime(new Date('2023-01-01T00:00:01.100Z'));
    log1.trace.begin({ id: 1, cat: 'parallel' }, 'duration1:stacked');

    jest.setSystemTime(new Date('2023-01-01T00:00:01.400Z'));
    log1.trace.end({ id: 1, cat: 'parallel' }, 'duration1:stacked');

    jest.setSystemTime(new Date('2023-01-01T00:00:01.500Z'));
    log1.trace.begin({ id: 2, cat: 'parallel' }, 'duration2');

    jest.setSystemTime(new Date('2023-01-01T00:00:02.000Z'));
    log1.trace.end({ id: 1, cat: 'parallel' });

    jest.setSystemTime(new Date('2023-01-01T00:00:02.500Z'));
    log1.trace.end({ id: 2, cat: 'parallel' });

    jest.setSystemTime(new Date('2023-01-01T00:00:00.500Z'));
    log2.info.begin({ cat: 'parallel' }, 'long-duration');

    jest.setSystemTime(new Date('2023-01-01T00:00:01.000Z'));
    log2.debug('instant event');

    jest.setSystemTime(new Date('2023-01-01T00:00:02.250Z'));
    log2.trace.end({ cat: 'parallel' });

    jest.useRealTimers();
    await Promise.all([root1.close(), root2.close()]);
  }
});
