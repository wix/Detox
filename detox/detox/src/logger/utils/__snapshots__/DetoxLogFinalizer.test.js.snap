// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetoxLogFinalizer createEventStream should create a stream of Chrome Trace format events 1`] = `
Array [
  Object {
    "args": Object {
      "name": "primary",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "sort_index": 0,
    },
    "name": "process_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "name": "main",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "sort_index": 0,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "level": 30,
      "v": 0,
    },
    "cat": "main",
    "name": "msg1",
    "ph": "i",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "name": "secondary",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "sort_index": 1,
    },
    "name": "process_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "level": 30,
      "v": 0,
    },
    "cat": "parallel",
    "name": "long-duration",
    "ph": "B",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration1",
    "ph": "B",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "level": 20,
      "v": 0,
    },
    "cat": "undefined",
    "name": "instant event",
    "ph": "i",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration1:stacked",
    "ph": "B",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531201100000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531201400000,
  },
  Object {
    "args": Object {
      "name": "parallel",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "sort_index": 1,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "id": 2,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration2",
    "ph": "B",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531202000000,
  },
  Object {
    "args": Object {
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531202250000,
  },
  Object {
    "args": Object {
      "id": 2,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531202500000,
  },
]
`;

exports[`DetoxLogFinalizer finalize should convert JSONL logs to Chrome Trace format: chrome-trace 1`] = `
Array [
  Object {
    "args": Object {
      "name": "primary",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "sort_index": 0,
    },
    "name": "process_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "name": "main",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "sort_index": 0,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "level": 30,
      "v": 0,
    },
    "cat": "main",
    "name": "msg1",
    "ph": "i",
    "pid": 1234,
    "tid": 0,
    "ts": 1672531200000000,
  },
  Object {
    "args": Object {
      "name": "secondary",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 1235,
    "tid": 0,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "sort_index": 1,
    },
    "name": "process_sort_index",
    "ph": "M",
    "pid": 1235,
    "tid": 0,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "name": "parallel",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1235,
    "tid": 3,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "sort_index": 3,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1235,
    "tid": 3,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "level": 30,
      "v": 0,
    },
    "cat": "parallel",
    "name": "long-duration",
    "ph": "B",
    "pid": 1235,
    "tid": 3,
    "ts": 1672531200500000,
  },
  Object {
    "args": Object {
      "name": "parallel",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "sort_index": 1,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration1",
    "ph": "B",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "name": "undefined",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1235,
    "tid": 4,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "sort_index": 4,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1235,
    "tid": 4,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "level": 20,
      "v": 0,
    },
    "cat": "undefined",
    "name": "instant event",
    "ph": "i",
    "pid": 1235,
    "tid": 4,
    "ts": 1672531201000000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration1:stacked",
    "ph": "B",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201100000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531201400000,
  },
  Object {
    "args": Object {
      "name": "parallel",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1234,
    "tid": 2,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "sort_index": 2,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1234,
    "tid": 2,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "id": 2,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "name": "duration2",
    "ph": "B",
    "pid": 1234,
    "tid": 2,
    "ts": 1672531201500000,
  },
  Object {
    "args": Object {
      "id": 1,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 1,
    "ts": 1672531202000000,
  },
  Object {
    "args": Object {
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1235,
    "tid": 3,
    "ts": 1672531202250000,
  },
  Object {
    "args": Object {
      "id": 2,
      "level": 10,
      "v": 0,
    },
    "cat": "parallel",
    "ph": "E",
    "pid": 1234,
    "tid": 2,
    "ts": 1672531202500000,
  },
]
`;

exports[`DetoxLogFinalizer finalize should convert JSONL logs to Chrome Trace format: plain 1`] = `
"2023-01-01T00:00:00.000Z detox[1234] i main msg1
2023-01-01T00:00:00.500Z detox[1235] B parallel long-duration
2023-01-01T00:00:01.000Z detox[1234] B parallel@1 duration1
2023-01-01T00:00:01.000Z detox[1235] i  instant event
2023-01-01T00:00:01.100Z detox[1234] B parallel@1 duration1:stacked
2023-01-01T00:00:01.400Z detox[1234] E parallel@1 duration1:stacked
2023-01-01T00:00:01.500Z detox[1234] B parallel@2 duration2
2023-01-01T00:00:02.000Z detox[1234] E parallel@1 duration1
2023-01-01T00:00:02.250Z detox[1235] E parallel long-duration
2023-01-01T00:00:02.500Z detox[1234] E parallel@2 duration2
"
`;
